### ISSUE TITLE
How to reproduce IS and FID in paper?

### ISSUE THREAD

**Issue by Jiang-Stan** [ThisisBillhe/PTQD#1]
Hi, thanks for your excellent work.
I have finished quantization following steps mentioned in readme, and successfully generate 50000 samples(saved in a npz file) for evaluation using quant_scripts/generate_samples_4_evaluation_brecq. How could I test these samples? It seems that I need to provide real data for evaluation by FID, where can I generate these data?

**Comment by ThisisBillhe** [comment]
Hi Jiang! First, you may want to use main_brecq_imagenet.py to see whether the model is quantized correctly. For evaluation, the real dataset is needed and [pytorch-fid ](https://github.com/mseitzer/pytorch-fid) can be used to calculate the FID.

**Comment by Jiang-Stan** [comment]
Thanks for your reply!
As far as I know, FID is calculated by μ and sigma of real dataset and generated dataset. Is there any "official real dataset" for comparison with generated images? Can I directly using mean and std of datasets such as ImageNet for calculating rather than go through all images in such dataset?

**Comment by ThisisBillhe** [comment]
The mean and std used to calculate FID are the statistics of the featuremaps of the inceptionV3 model, not the source dataset. There are some pre-calculated reference batch [here](https://github.com/openai/guided-diffusion/tree/main/evaluations).

**Comment by Jiang-Stan** [comment]
Got it. Thanks for your sincerely help!

**Comment by Jiang-Stan** [comment]
Hi, sorry for bothering you the second time, but I'm really confused...
I tested IS using [this repo](https://github.com/toshas/torch-fidelity), which is installed through environment.yaml. Images(50k) generated by 4w8f model performes too well to believe in(IS=299),  far surpassing the IS=211 tested on ImageNet val data(50k also). Is there any official reports for validating correctness of IS? Or did you use other tools for calculating IS?
Thank you!

**Comment by ThisisBillhe** [comment]
IS and FID are highly related to ddim_steps/eta/scales, what are these hyperparameters you use? Also, there are many papers with IS results higher than 211, such as LDM and DiT. So I think it's not a big deal as long as the generated images look fine.  I use [this repo](https://github.com/openai/guided-diffusion/tree/main/evaluations) for calculating IS, which requires TensorFlow.

**Comment by Jiang-Stan** [comment]
I found your description about evaluation in paper just now... Thank you very much!

**Comment by Jiang-Stan** [comment]
I tried to reproduce your result with step=20,eta=0.0,scale=3.0, without modifying your code, and evaluate it using evaluation code in guided_diffusion. The result is shown as below:
![2023-07-07 16-42-13 的屏幕截图](https://github.com/ThisisBillhe/PTQD/assets/74638604/4c4f5e5e-8908-49b8-8e9d-79101bb22599)
The sFID seems close to paper reported, but IS and FID significantly changed.
I noticed that IS and FID of 8w8f are worse than those of 4w8f, as shown in your paper:
![2023-07-07 16-35-59 的屏幕截图](https://github.com/ThisisBillhe/PTQD/assets/74638604/02424be1-d400-46e1-8550-7cc4c67d0c97)
Does it means that sFID is important than other two metrics? And is it a common phenomenon that IS and FID would change sharply once the generated images changed?

**Comment by ThisisBillhe** [comment]
> I tried to reproduce your result with step=20,eta=0.0,scale=3.0, without modifying your code, and evaluate it using evaluation code in guided_diffusion. The result is shown as below: ![2023-07-07 16-42-13 的屏幕截图](https://user-images.githubusercontent.com/74638604/251688221-4c4f5e5e-8908-49b8-8e9d-79101bb22599.png) The FID/sFID seems close to paper reported, but IS significantly changed. I noticed that IS and FID of 8w8f are worse than those of 4w8f, as shown in your paper: ![2023-07-07 16-35-59 的屏幕截图](https://user-images.githubusercontent.com/74638604/251686215-02424be1-d400-46e1-8550-7cc4c67d0c97.png) Does it means that sFID is important than other two metrics? And is it a common phenomenon that IS and FID would change sharply once the generated images changed?

I think researchers pay more attention to FID/sFID and many papers only report on these. Some implementation details in this repo may differ from the paper, especially BRECQ's settings.  In brecq_quant_model.py, Setting module_list[1] to 8-bit (which is indeed an input layer for images) while uncommenting lines for module_list[2] might help to improve performance.

**Comment by Jiang-Stan** [comment]
Hi He, I found that the [code here](https://github.com/ThisisBillhe/PTQD/blob/master/quant_scripts/generate_samples_4_evaluation_brecq.py#L189) might be the reason of worse performance. By directly multiply it with 255. and uniformally sampling 50 images from each class, I got better tested performance using float model, as shown below:
![2023-07-11 18-30-31 的屏幕截图](https://github.com/ThisisBillhe/PTQD/assets/74638604/28f12088-d628-4f90-9e26-d17a66cc9153)

**Comment by ThisisBillhe** [comment]
Thanks for your notice!! (I forgot to reply

**Comment by JiaojiaoYe1994** [comment]
Hi, thank you for your cool work. I have a question regarding Bit Operations (BOPs) measure. How do you implement this metric?

**Comment by TaylorJocelyn** [comment]
> Hi He, I found that the [code here](https://github.com/ThisisBillhe/PTQD/blob/master/quant_scripts/generate_samples_4_evaluation_brecq.py#L189) might be the reason of worse performance(with which I got FID/sFID>17 using imgs generated by float model). By directly multiply it with 255(as done in [latent-diffusion](https://colab.research.google.com/github/CompVis/latent-diffusion/blob/main/scripts/latent_imagenet_diffusion.ipynb)). and uniformally sampling 50 images from each class, I got better tested performance using float model, as shown below: ![2023-07-11 18-30-31 的屏幕截图](https://private-user-images.githubusercontent.com/74638604/252636279-28f12088-d628-4f90-9e26-d17a66cc9153.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYxMTU0MzAsIm5iZiI6MTcxNjExNTEzMCwicGF0aCI6Ii83NDYzODYwNC8yNTI2MzYyNzktMjhmMTIwODgtZDYyOC00ZjkwLTllMjYtZDE3YTY2Y2M5MTUzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE5VDEwMzg1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMxOWFmZDhhMWFjYmUyMTQxNGJiNWJlMGQ0YTgwNzQ3MTlkOGRhMDUwNWU4OTM2NzY1OWQ2OGFjZWViNGIwZGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.QpKQrg2bB_dD-Uc7a91bCLSiU0VqyjPhdTcVCbuurQo)
> 
> And I successfully reproduced results reported by latent-diffusion in this way. As shown below(IS by torch-fidelity, others by guided_diffusion evaluation code. The paper claims that FID calculated by both tools is almost coincide):
> 
> cfg with step=250, scale=1.5	IS↑	FID↓	sFID↓	Prec.↑	Recall↑
> Paper reported	247.67±5.59	3.60	-	87%	48%
> eta=0	205.55±5.27	3.31	5.10	82.95%	53.57%
> eta=1	249.59±3.30	3.54	5.10	87.15%	48.50%

@Jiang-Stan Congratulations on successfully solving the problem you encountered！
 Seeing that you have achieved better results，I would like to ask you a few question.
1. In collect_diffusion_input_imagenet.py file, how did you set classes and n_samples_per_class?
<img width="1207" alt="image" src="https://github.com/ziplab/PTQD/assets/60263416/24e319d6-ff7d-412d-a17f-3a6ac082e584">
2. In collect_quant_error_brecq.py file, how did you set classes and n_samples_per_class?
<img width="1032" alt="image" src="https://github.com/ziplab/PTQD/assets/60263416/773d61cc-b791-4156-9035-0b5e7c74f930">
Looking forward to your answer. Thank you very much.
